src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseService.java

```diff
@@ -646,14 +648,16 @@ public void commitAndPushRepository(Repository repository, String message, User
      */
     public ProgrammingExercise updateTimeline(ProgrammingExercise updatedProgrammingExercise, @Nullable String notificationText) {
         var programmingExercise = programmingExerciseRepository.findByIdElseThrow(updatedProgrammingExercise.getId());
+        final ProgrammingExercise programmingExerciseBeforeUpdate = programmingExercise;
```

Comment:
Yes, I think you are correct. I tried some experiments in an online compiler with objects and sadly it seems to be the case that the variable even though it is final still gets changed due to the referencing. 
I added a new version of the method with a bool at the start and check the release dates at the beginning therefore I no longer need a copy of the object. Btw if you know an easy way to copy the exercise by value feel free to share because I could not find one ( copy() is not defined and a second db call seems to be quite redundant)

Reply:
There is Object.clone() and Cloneable, but that would require changing the exercise class for questionable benefits.
Next option would be to just create a new object and just set the release date, as this is the only thing actually required for us. But then we would have an Exercise object going around that pretends to be the whole thing but actually only one property is valid.
But instead we can just pass the release date by itself and not worry about the rest of the exercise
suggested changes in `bugfix/notification-system/release-notification-change-01` 
Not sure if you want to just pass the date when you have a fully exercise available, it left those as is to reduce the number of changes.
also contains a fix for the other issue.

Reply:
I decided to simply create a "slim copy" of the exercise and copy the release date.